feat: release automation script

IronCore864 opened the pull request.

dimaqq commented on release.py:
```diff
@@ -0,0 +1,498 @@
+
+    def format(self, record: logging.LogRecord):
+        """Format the log record based on its level."""
+        if record.levelno == logging.INFO:
+            self.setFormatter(logging.Formatter('ℹ️ INFO: %(message)s'))  # noqa: RUF001
+        elif record.levelno == logging.WARNING:
+            self.setFormatter(logging.Formatter('⚠️ WARNING: %(message)s'))
+        elif record.levelno == logging.ERROR:
+            self.setFormatter(logging.Formatter('❌ ERROR: %(message)s'))
+        return super().format(record)
```
That's a very roundabout way of doing this.

The handler's formatter is bounced around on each record.

If you are determined to go this way, there must be an `else:` clause to support other, unlisted log levels.

I would recommend setting a **custom formatter** on the **stock handler** instead.

Here's one example: https://github.com/dimaqq/python-libjuju/blob/cd5930c85d0f8bf5417bc9394d2c767ecc6e5e85/example.py#L63-L93

dimaqq commented on release.py:
```diff
@@ -0,0 +1,498 @@
+    logger.critical('Environment variable GITHUB_TOKEN not set.')
+    exit(1)
+
+OPS_VERSION_STR = r'version: str = \'(\d+\.\d+\.\d+(?:\.dev\d+)?)\''
```
I would recommend start and end of line markers to be more precise.

```suggestion
OPS_VERSION_STR = r'^version: str = \'(\d+\.\d+\.\d+(?:\.dev\d+)?)\'$'
```

`re.xxx` calls may need the `re.MULTILINE` flag.

Without these markers, other matches are possible, for example:

```py
testing/src/scenario/_runtime.py
62:        juju_version: str = '3.0.0',
```

dimaqq commented on release.py:
```diff
@@ -0,0 +1,498 @@
+
+    new_branch = f'release-prep-{tag}'
+    subprocess.run(['/usr/bin/git', 'checkout', '-b', new_branch], check=True)
+    changed_files = ['CHANGES.md', *VERSION_FILES]
+    for file in changed_files:
+        subprocess.run(['/usr/bin/git', 'add', file], check=True)
```
I think that we need to validate that each file was in fact changed.
`git add` returns OK status even if there are no changes.

Perhaps it's better to fail early in the `update_versions_**` or `update_ops/pyproject_version` functions.

dimaqq commented on release.py:
```diff
@@ -0,0 +1,498 @@
+
+    update_versions_for_post_release()
+
+    new_branch = 'post-release'
+    subprocess.run(['/usr/bin/git', 'checkout', '-b', new_branch], check=True)
+    for file in VERSION_FILES:
+        subprocess.run(['/usr/bin/git', 'add', file], check=True)
+    subprocess.run(['/usr/bin/git', 'commit', '-m', 'chore: post release'], check=True)
+    subprocess.run(['/usr/bin/git', 'push', 'origin', new_branch], check=True)
+    pr = repo.create_pull(
+        title='Post Release',
+        body='This PR updates the version files after the release.',
+        head=f'{gh_client.get_user().login}:{new_branch}',  # "your_username:new_branch"
```
Corner cases:
- `post-release` branch exists locally
- `post-release` branch exists on GitHub but not locally

Perhaps the script could take a stance here--either fail, on `check=True` (current behaviour) or overwrite local and remote branches with this name, or generate a random name, like `f"post-release-{some_hex}"` (maybe secrets.token_hex() or in fact current git commit)

dimaqq commented on uv.lock:
```diff
@@ -1,5 +1,5 @@
 version = 1
-revision = 1
+revision = 2
```
**Off-topic:** we all need to upgrade our `uv`

dimaqq replied:
Ah, nevermind.,
I get the same lock version bump if I remove the file and re-lock.
It seems that uv keeps current format if it can, and the only difference between the formats is the addition of the PYPI timestamp.

dimaqq reviewed (changes requested):
Early review

tonyandrewmeyer commented on release.py:
```diff
@@ -0,0 +1,848 @@
+    )
+    subprocess.run(['/usr/bin/git', 'push', 'origin', new_branch], check=True)
+    pr = repo.create_pull(
+        title='Post Release',
```
This will have to be something like `chore: adjust versions after release` so that it fits with the conventional commits.

james-garner-canonical commented on release.py:
```diff
@@ -0,0 +1,848 @@
+class SymbolFormatter(logging.Formatter):
+    """Custom logging formatter to use symbols for log levels."""
+
+    from typing import ClassVar
+
+    levels: ClassVar[dict[str, str]] = {
```
My suggestion here would be to just define this in the module scope (e.g. `_LOG_LEVELS = {...`), no type annotation required.

If you want to keep it in the class body, then I think that using `Mapping[str, str]` would be cleaner than the `ClassVar`. The `from typing import ...` line should go at the top of the script with the other imports.

tonyandrewmeyer commented on release.py:
```diff
@@ -0,0 +1,848 @@
+    for line in release_notes.splitlines():
+        line = line.strip()
+        if line.startswith('* '):
+            match = re.match(r'\* (\w+): (.*) by @\w+ in (.*)', line)
```
I think we want to keep (in the release notes, not so much the changelog) acknowledgement of external contributions, don't we?

IronCore864 replied:
Per [discussion here](https://docs.google.com/document/d/1fkoqp1A2SMiOnwHubOJVRkuD1GpvcYxSDmfExYIZO88/edit?tab=t.0) (internal link), we have agreed on "Drop the GH usernames (even if they come from outside the team)."

tonyandrewmeyer commented on release.py:
```diff
@@ -0,0 +1,848 @@
+            logger.info(
+                f'Note that there are {len(open_prs) - 1} more open update charm pins PRs.'
+            )
+        countdown('Exiting and opening the charm pins PR in: ', 3)
+        webbrowser.open(pr.html_url)
```
Also here, I would rather just click the link. But if everyone else likes their browser being opened then I can live with that.

james-garner-canonical replied:
+1 for a link instead, please

tonyandrewmeyer commented on release.py:
```diff
@@ -0,0 +1,848 @@
+    return new_version
+
+
+def get_new_scenario_version_post_release() -> str:
```
I would be fine if we removed this and made it `{ops_major+5}.{ops_minor}.{ops.bugfix}`.

Or if we are ok with ops-scenario bumping a major version without ops, then I think we should handle that for ops-tracing too. But I think we are sticking with no major version changes for now?

IronCore864 replied:
If there is no major version change, it is fine to use a suggested version, either from `{ops_major+5}.{ops_minor}.{ops.bugfix}`, or parse what's in the `testing/pyproject.toml` file and bump the minor version (part of the logic in this function).

Since this script was created during the 2.x -> 3 release, I added an optional override. There will still be an automatically suggested version in case you don't want to input anything, so I am keeping this logic for now.

tonyandrewmeyer commented on release.py:
```diff
@@ -0,0 +1,848 @@
+) -> github.GitRelease.GitRelease | None:
+    """Create a draft release with auto-generated notes."""
+    max_retries = 3
+    for attempt in range(1, max_retries + 1):
```
Why is this needed? We don't have the retries everywhere else - is creating a release unreliable?

IronCore864 replied:
I added this because in my network environment, I met failures here. I added retries elsewhere, making the script much more complicated, so I reverted most changes and kept only this one. But I agree it should be consistent, so I will remove this one too.

tonyandrewmeyer reviewed (commented):
A few comments, but in general I think the primary test is a release, and the last two, which used this, seemed fine, and if we think there should be tweaks as we each use this to do a release, doing it at the time seems fine.

james-garner-canonical commented on release.py:
```diff
@@ -0,0 +1,848 @@
+            match = re.match(r'\* (\w+): (.*) by @\w+ in (.*)', line)
+            if match:
+                category = match.group(1)
+                description = match.group(2).strip()
+                description = description[0].upper() + description[1:]
```
```suggestion
                description = match.group(2).strip().capitalize()
```

james-garner-canonical commented on release.py:
```diff
@@ -0,0 +1,848 @@
+                description = match.group(2).strip()
+                description = description[0].upper() + description[1:]
+                pr_link = match.group(3).strip()
+                if category in categories:
+                    categories[category].append((description, pr_link))
```
Suggestion: put the `definition` and `pr_link` creation in here too.

IronCore864 replied:
The reason behind my current implementation is:

- `parse_release_notes` only parses PRs into categories, then the result is used later
- When creating a release note, use the result from `parse_release_notes`, and format it as to release note (`format_release_notes`).
- When updating `CHANGES.md`, use the result from `parse_release_notes`, and format it accordingly; that's why the description and PR links are handled in `format_changes`.

james-garner-canonical commented on release.py:
```diff
@@ -0,0 +1,848 @@
+
+def update_changes_file(changes: str, file: str):
+    """Update the changes file with new release notes."""
+    existing_content = ''
+    if os.path.exists(file):
+        with open(file) as f:
+            existing_content = f.read()
+
+    new_content = changes + existing_content
+
+    with open(file, 'w') as f:
+        f.write(new_content)
+
```
We should probably use `pathlib` for path stuff. Then this could be, e.g.:
```suggestion
    existing_content = path.read_text() if path.exists() else ''
    path.write_text(changes + existing_content)
```

james-garner-canonical commented on release.py:
```diff
@@ -0,0 +1,848 @@
+    file = VERSION_FILES['testing']
+    file_path = Path(file)
+    content = file_path.read_text()
+    match = re.search(r'version = "' + VERSION_REGEX + '"', content)
```
Does a raw f-string work here?
```suggestion
    match = re.search(rf'version = "{VERSION_REGEX}"', content)
```

james-garner-canonical commented on release.py:
```diff
@@ -0,0 +1,848 @@
+
+
+VERSION_REGEX = r'(\d+\.\d+\.\d+(?:\.dev\d+)?)'
+VERSION_FILES = {
+    'ops/src': 'ops/version.py',
+    'ops/pyproject': 'pyproject.toml',
+    'testing': 'testing/pyproject.toml',
+    'tracing': 'tracing/pyproject.toml',
+    'uvlock': 'uv.lock',
```
Speaking of `pathlib`, maybe it would pay to make these `pathlib.Path` where they're defined?
```suggestion
VERSION_FILES = {
    'ops/src': Path('ops/version.py'),
    'ops/pyproject': Path('pyproject.toml'),
    'testing': Path('testing/pyproject.toml'),
    'tracing': Path('tracing/pyproject.toml'),
    'uvlock': Path('uv.lock'),
```

james-garner-canonical commented on release.py:
```diff
@@ -0,0 +1,848 @@
+    content = file_path.read_text()
+    updated = re.sub(
+        r'version = "' + VERSION_REGEX + '"',
+        f'version = "{testing_version}"',
+        content,
+    )
+    updated = re.sub(
+        r'ops==' + VERSION_REGEX,
+        f'ops=={ops_version}',
+        updated,
+    )
+    # Fail if nothing is changed.
+    if content == updated:
+        logger.error(f'No changes made to {file}. Check the versions.')
+        exit(1)
+    file_path.write_text(updated)
+    logger.info(f'Updated {file} to ops {ops_version} testing {testing_version}')
+
+
+def update_tracing_version(ops_version: str):
+    """Update the tracing pyproject version."""
+    file = VERSION_FILES['tracing']
+    file_path = Path(file)
+    content = file_path.read_text()
+    updated = re.sub(
+        r'version = "' + VERSION_REGEX + '"',
+        f'version = "{ops_version}"',
+        content,
+    )
+    updated = re.sub(
+        r'ops==' + VERSION_REGEX,
+        f'ops=={ops_version}',
+        updated,
+    )
+    # Fail if nothing is changed.
+    if content == updated:
+        logger.error(f'No changes made to {file}. Check the versions.')
+        exit(1)
+    file_path.write_text(updated)
+    logger.info(f'Updated {file} to version {ops_version}')
```
There seems to be a lot of duplication here that could perhaps be factored out cleanly before we merge this.

IronCore864 replied:
I had a draft without duplication in the first place, but it requires many if/elses, since:

- For each module, we need to decide if we are only updating a source file or also `pyproject.toml`
- For some modules, it only needs to update the version of itself, but for others, it also needs to update the ops version.

So I ended up with the current version, with duplicated code but much more readable.

dimaqq commented:
Re PR description:

> git command available locally

I imagine it's pretty typical though. We all use git.

The PR doesn't seem to shell out to `git` or `gh`.

Could it be that the PR description is out of date?

IronCore864 commented on release.py:
```diff
@@ -0,0 +1,848 @@
+
+
+class SymbolFormatter(logging.Formatter):
+    """Custom logging formatter to use symbols for log levels."""
```
In my testing and debugging, I found it hard to go through all those outputs, so I adopted this symbol formatter from [here](https://github.com/dimaqq/python-libjuju/blob/cd5930c85d0f8bf5417bc9394d2c767ecc6e5e85/example.py#L63-L93) based on Dima's suggestion. Now that the script is steady and requires less debugging, I am OK to drop this. Reverting it to a simple formatter, something like:

```python
formatter = logging.Formatter('%(asctime)s - %(levelname)s - %(message)s')

logger = logging.getLogger('release')
logger.setLevel(logging.DEBUG)

handler = logging.StreamHandler()
handler.setFormatter(formatter)

logger.addHandler(handler)
```

IronCore864 commented on release.py:
```diff
@@ -0,0 +1,848 @@
+        '--owner',
+        '-o',
+        help='Owner name (e.g. "Canonical")',
+        default='Canonical',
```
upper/lower: updated to lower.

Hard-code V.S. default values: when testing, I need to create a repo under `my-org/repo`, then fork it to `my-gh-handle/repo`, then make the release from the latter, so I did the defaults with a possibility to override for local testing.

IronCore864 commented on release.py:
```diff
@@ -0,0 +1,732 @@
+import github.Repository
+
+formatter = logging.Formatter('%(asctime)s - %(levelname)s - %(message)s')
+logger = logging.getLogger('release')
+logger.setLevel(logging.DEBUG)
+handler = logging.StreamHandler()
+handler.setFormatter(formatter)
```
Yes, I will update it to INFO. Thanks for pointing this out.
I had some debugging stuff, but I don't think they are extremely helpful, so I removed them.

IronCore864 commented:
Another test after refactoring according to all comments above:

```bash
$ git branch
  main

$ python release.py --owner ironcoreworks --repo release-automation  # testing org and repo
2025-07-10 11:13:14,849 - INFO - Latest tag in branch "main": 0.1.0
2025-07-10 11:13:14,850 - INFO - Suggested new version: 0.2.0
Input the new tag for the release (press enter to use the tag 0.2.0), or 'c' to cancel:
3.2.0
2025-07-10 11:13:17,939 - WARNING - Check out the releases page: https://github.com/ironcoreworks/release-automation/releases before confirming!
Confirm creating tag '3.2.0' on branch 'main'? [y/N]: y
2025-07-10 11:13:19,675 - INFO - Draft release created: https://github.com/IronCoreWorks/release-automation/releases/tag/untagged-5a42876b7b5b428d3b6b.
================================================================================
Formatted release notes:
================================================================================
## What's Changed


### Features
* Add changes.md in https://github.com/IronCoreWorks/release-automation/pull/6
* Update release automation script and versions in https://github.com/IronCoreWorks/release-automation/pull/20
* Update in https://github.com/IronCoreWorks/release-automation/pull/23
* Update release script in https://github.com/IronCoreWorks/release-automation/pull/24

### Documentation
* Add version in https://github.com/IronCoreWorks/release-automation/pull/3

### Tests
* Add test in https://github.com/IronCoreWorks/release-automation/pull/4

**Full Changelog**: https://github.com/IronCoreWorks/release-automation/compare/0.1.0...3.2.0
================================================================================
2025-07-10 11:13:19,676 - INFO - The automatically generated title is: 3.2.0
Enter release title, press Enter to keep the auto-generated title:
> Test release, ops 3.2.0, tracing 3.2.0, testing/scenario 8.2.0

Enter release summary (multi-line supported; type '.' on a new line to finish):
This is a release for testing the automation script.

The previous tag from my repo is 0.1.0, hence the suggested version is 0.2.0, which is overriden by values provided by the user.

A quick brown fox jumps over something that I forgot.

Put a single dot at the final line to finish drafting the introduction paragraph.
.
2025-07-10 11:13:38,740 - INFO - Release updated: https://github.com/IronCoreWorks/release-automation/releases/tag/untagged-01c17942f2ee6832d913.
2025-07-10 11:13:38,742 - INFO - Updated CHANGES.md with new release notes.
2025-07-10 11:13:38,742 - INFO - Suggested new scenario version: 8.2.0
Input new scenario version (press enter to use the suggested version 8.2.0), or 'c' to cancel:
Confirm using scenario version '8.2.0'? [y/N]: y
2025-07-10 11:13:47,429 - INFO - Updated ops/version.py to version 3.2.0
2025-07-10 11:13:47,430 - INFO - Updated pyproject.toml to version 8.2.0
2025-07-10 11:13:47,430 - INFO - Updated testing/pyproject.toml to version 8.2.0
2025-07-10 11:13:47,431 - INFO - Updated tracing/pyproject.toml to version 3.2.0
Using CPython 3.12.9
Resolved 149 packages in 304ms
Updated ops-scenario v8.2.0.dev0 -> v8.2.0
Updated ops-tracing v3.2.0.dev0 -> v3.2.0
Switched to a new branch 'release-prep-3.2.0'
[release-prep-3.2.0 0a8f4d0] chore: prepare release 3.2.0
 6 files changed, 26 insertions(+), 9 deletions(-)
Enumerating objects: 20, done.
Counting objects: 100% (20/20), done.
Delta compression using up to 10 threads
Compressing objects: 100% (8/8), done.
Writing objects: 100% (11/11), 2.03 KiB | 2.03 MiB/s, done.
Total 11 (delta 6), reused 0 (delta 0), pack-reused 0
remote: Resolving deltas: 100% (6/6), completed with 6 local objects.
remote:
remote: Create a pull request for 'release-prep-3.2.0' on GitHub by visiting:
remote:      https://github.com/IronCore864/release-automation/pull/new/release-prep-3.2.0
remote:
To github.com:IronCore864/release-automation.git
 * [new branch]      release-prep-3.2.0 -> release-prep-3.2.0
2025-07-10 11:13:53,908 - INFO - Created PR: https://github.com/IronCoreWorks/release-automation/pull/26
2025-07-10 11:13:53,908 - INFO - Draft release created. Please merge the version bump PR, publish the draft release, thenrun this script with --post-release to perform post-release actions.
```

tonyandrewmeyer commented on release.py:
```diff
@@ -0,0 +1,694 @@
+    org = gh_client.get_organization(owner)
+    repo = org.get_repo(repo_name)
+
+    check_update_charm_pins_prs(repo)
```
I wonder if we might need an override for this, but I think it's fine to leave as-is for now, and we can add that if it does come up in the future.

tonyandrewmeyer reviewed (changes requested):
I'm happy with this as-is. I'm sure we'll want to tweak things in the future, but this seems a good starting point. It's only for our use, so we have more flexibility than with most changes.

The one change I'd like is for HACKING.md to be updated to include instructions for running this script and remove instructions for all the things this script does.

benhoyt commented on release.py:
```diff
@@ -0,0 +1,694 @@
+import os
+import re
+import subprocess
+from datetime import datetime
+from pathlib import Path
```
[Our style](https://github.com/canonical/operator/blob/main/STYLE.md#import-modules-not-objects) is `import package`:
```suggestion
import datetime
import pathlib
```

benhoyt commented on release.py:
```diff
@@ -0,0 +1,694 @@
+
+def get_latest_release_tag(repo: github.Repository.Repository, branch_name: str) -> str | None:
+    """Get the latest release tag from the repository."""
+    releases = repo.get_releases()
```
Are we certain this returns the releases in newest-first order?

benhoyt replied:
Yeah, if it's not guaranteed/specified in the docs, I think we should do a quick `.sort()` by `published_at`, to avoid surprises. Then again, they'll probably never change this ... it'd break too many people, so maybe it's okay.

benhoyt replied:
Sounds reasonable to keep it as is, thanks James.

benhoyt commented on release.py:
```diff
@@ -0,0 +1,694 @@
+def bump_minor_version(version: str) -> str:
+    """Bump minor version."""
+    major, minor, patch = map(int, version.split('.'))
+    return f'{major}.{minor + 1}.{patch}'
```
Would we want to reset the patch version to 0 here? It would seem odd to bump from say 2.23.4 to 2.24.4.

benhoyt commented on release.py:
```diff
@@ -0,0 +1,694 @@
+            logger.info(f'Suggested new version: {suggested_tag}')
+
+    tag_prompt = f' (press enter to use the tag {suggested_tag})' if suggested_tag else ''
+    prompt = f"Input the new tag for the release{tag_prompt}, or 'c' to cancel:\n"
```
I don't think we need special handling of "c" to cancel. It's common practice to press Ctrl-C to cancel these kind of scripts (which Python would handle automatically).

benhoyt commented on release.py:
```diff
@@ -0,0 +1,694 @@
+            logger.warning('Release canceled.')
+            exit(0)
+
+        new_tag = user_input if user_input else suggested_tag
```
Nit, equivalent to:
```suggestion
        new_tag = user_input or suggested_tag
```

benhoyt commented on release.py:
```diff
@@ -0,0 +1,694 @@
+        logger.warning(f'Check out the releases page: {release_page} before confirming!')
+
+        confirm = (
+            input(f"Confirm creating tag '{new_tag}' on branch '{branch_name}'? [y/N]: ")
```
You can get rid of all the extra quotes and have Python add them for you by using the `!r` format (repr). Similar elsewhere:

```suggestion
            input(f'Confirm creating tag {new_tag!r} on branch {branch_name!r}? [y/N]: ')
```

benhoyt commented on release.py:
```diff
@@ -0,0 +1,694 @@
+            target_commitish=branch,
+        )
+        return release
+    except Exception as e:
```
I would prefer to simplify and just to let this exception bubble up and exit. We're exiting anyway, and it avoids the try/except boilerplate here and the check in the caller, and provides a full stack trace for the developer using the tool.

benhoyt commented on release.py:
```diff
@@ -0,0 +1,694 @@
+    for line in release_notes.splitlines():
+        if match := re.match(r'^\* (\w+): (.*) by @\w+ in (.*)', line.strip()):
+            category = match.group(1)
+            if category in categories:
```
Should we have an `else` and print a log message if we run into a category we don't expect? Or is that not an issue?

benhoyt commented on release.py:
```diff
@@ -0,0 +1,694 @@
+                    pr_num = match.group(1)
+                res += f'* {description} (#{pr_num})\n'
+
+            res += '\n'
```
Let's use a similar approach to `format_release_notes` with a `lines` list and `'\n'.join(lines)` at the end.

benhoyt commented on release.py:
```diff
@@ -0,0 +1,694 @@
+            f"{suggested_version or 'required'}), or 'c' to cancel: "
+        ).strip()
+
+        if user_input.lower() == 'c':
```
Similar here -- I think we should drop this 'c' handling and just let the user press Ctrl-C.

benhoyt commented on release.py:
```diff
@@ -0,0 +1,694 @@
+    return base_version, dev_suffix
+
+
+def get_new_scenario_version_for_release() -> str:
```
Can we skip/automate this step entirely? Don't we want the scenario version to always be exactly major+5, like 3.1.2 -> 8.1.2? (For now at least, and we can always adjust the code later if needed.)

IronCore864 replied:
We cannot skip it all, because the scenario version differs from the ops version.

This function parses the scenario version from `pyproject.toml` and removes the `.dev0` suffix, uses it as a suggested scenario version for the release, and if the user doesn't accept, they can override it by providing a manually input version for scenario. The logic isn't major +5.

benhoyt commented on release.py:
```diff
@@ -0,0 +1,694 @@
+    return new_version
+
+
+def get_new_scenario_version_post_release() -> str:
```
Similar here, can we completely automate this for the scenario version?

benhoyt commented on release.py:
```diff
@@ -0,0 +1,694 @@
+    update_uv_lock()
+
+
+def get_new_version_post_release(repo: github.Repository.Repository, branch_name: str) -> str:
```
Can we completely automate this? Won't this always be minor+1 with the `.dev` suffix?

IronCore864 replied:
This exists specifically for major releases (I made the script during 2.23 -> 3.0 release). If the previous version is 2.23, and now we want to release 3.0, a manual override is required. But I don't think we will make any major release soon, so I will fully automate this.

benhoyt commented on release.py:
```diff
@@ -0,0 +1,694 @@
+        if len(open_prs) > 1:
+            logger.info(
+                f'Note that there are {len(open_prs) - 1} more open update charm pins PRs.'
+            )
```
Don't we want to exit at this point, and they re-run the script when they've merged them? (Or at least pause and wait for them to press enter or something.)

benhoyt commented on release.py:
```diff
@@ -0,0 +1,694 @@
+    new_branch = f'release-prep-{tag}'
+    subprocess.run(['/usr/bin/git', 'checkout', '-b', new_branch], check=True)
+    changed_files = ['CHANGES.md', *[str(path) for path in VERSION_FILES.values()]]
+    for file in changed_files:
+        subprocess.run(['/usr/bin/git', 'add', file], check=True)
```
Could do it in one operation instead, something like:
```suggestion
    subprocess.run(['/usr/bin/git', 'add', *changed_files], check=True)
```

benhoyt commented on release.py:
```diff
@@ -0,0 +1,694 @@
+        )
+        exit(1)
+
+    subprocess.run(['/usr/bin/git', 'fetch', 'upstream'], check=True)
```
These days my `canonical/operator` remote is called `origin` (and `benhoyt` is mine), not `upstream`. Can we avoid hard-coding this?

Could we do `git checkout {base_branch}` and `git pull` instead? And do we need to do the same "update to latest" before `draft_release` too?

IronCore864 replied:
`base_branch` defaults to `main`, so `git checkout main` and `git pull` will get the forked repo's main, instead of `canonical/operator`'s main. I will make the remote name configurable.

IronCore864 replied:
Before `draft_release`: I assume `tox -e draft-release` is executed in the forked ops repo, checked out at latest main, with no changes in the staging area. The reason I did not do any operations before `draft_release` is that in case you are working on something, it might ruin the result. I will add some description in HACKING.md.

benhoyt commented on release.py:
```diff
@@ -0,0 +1,694 @@
+
+    subprocess.run(['/usr/bin/git', 'checkout', '-b', new_branch], check=True)
+    for file in [str(path) for path in VERSION_FILES.values()]:
+        subprocess.run(['/usr/bin/git', 'add', file], check=True)
```
Similar here, let's do this in one `git add` operation with `*files`.

benhoyt commented on release.py:
```diff
@@ -0,0 +1,694 @@
+
+    draft_release(owner=args.owner, repo_name=args.repo, base_branch=args.branch)
+    logger.info(
+        'Draft release created. Please merge the version bump PR, publish the draft release, then'
```
```suggestion
        'Draft release created. Please merge the version bump PR, review and publish the draft release, then'
```

benhoyt reviewed (changes requested):
Looking close. A few changes requested (and I agree about updating HACKING.md). Let's try to get this merged by the end of this week.

benhoyt reviewed (approved):
Looks good, thanks for the updates. Approving optimistically, but a few minor changes to HACKING.md suggested.

IronCore864 commented on HACKING.md:
```diff
@@ -234,63 +234,38 @@ the build frontend is [build](https://pypi.org/project/build/).
+
+    > This assumes a draft release on the main branch, and your forked remote name is `origin`, and the `canonical/operator` remote name is `upstream`.
+    > 
+    > If you have different settings, add parameters accordingly. For example, the following command assumes your forked remote name is `mine`, and `canonical/operator` remote name is `origin`:
```
Per discussion, we agreed to use origin as the remote name of the forked repo, and upstream for canonical/operator. I have already refactored the script so that they are not hard-coded, and default values are provided.

tonyandrewmeyer commented on HACKING.md:
```diff
@@ -234,63 +234,41 @@ the build frontend is [build](https://pypi.org/project/build/).
+    > 
+    > `tox -e draft-release -- --canonical-remote origin --fork-remote mine`
+    > 
+    > If you want to make a release on another branch, for example, on "2.23-maintenance", run:
```
Do I still run the command with my local clone in the main branch? Or do I switch/checkout the branch I want to release first? Or will the script take care of it and it doesn't matter?

tonyandrewmeyer commented on HACKING.md:
```diff
@@ -234,63 +234,41 @@ the build frontend is [build](https://pypi.org/project/build/).
+    > [ops-tracing](https://pypi.org/project/ops-tracing/)).
+    > Note that it sometimes take a bit of time for the new releases to show up.
+    > 
+    > See [.github/workflows/publish-ops.yaml](.github/workflows/publish-ops.yaml) and
+    > [.github/workflows/publish-ops-scenario.yaml](.github/workflows/publish-ops-scenario.yaml) for details.
```
@dimaqq changed this underneath you :smile:. It's just one workflow now.

tonyandrewmeyer reviewed (approved):
A couple of small notes, but looks great, thanks!

IronCore864 commented:
Merged main into this branch, resolved the conflicts in HACKING.md, also tweaked the description in HACKING.md a little bit. Merging now.

IronCore864 commented on release.py:
```diff
@@ -0,0 +1,609 @@
+
+def get_latest_release_tag(repo: github.Repository.Repository, branch_name: str) -> str | None:
+    """Get the latest release tag from the repository."""
+    releases = repo.get_releases()
```
As mentioned in another comment, although not officially stated, the return result is sorted by release date (unless there is a draft release). Also, even if this function doesn't find anything, the user can still provide an override. So I'm going to keep this as it is.

IronCore864 commented on release.py:
```diff
@@ -0,0 +1,609 @@
+}
+
+
+def get_latest_release_tag(repo: github.Repository.Repository, branch_name: str) -> str | None:
```
Both should work, and they should be the same. I think we should do both, do a comparison, if not the same, ask the user to override. But probably it's not necessary, since the return value is only used as a suggested next release version, which can be overriden. So I'm keeping it as it is for now.

dimaqq commented on release.py:
```diff
@@ -0,0 +1,607 @@
+    Results in a Markdown formatted string with sections for each commit type.
+    If `full_changelog` is provided, it is appended at the end.
+    """
+    lines = ["## What's Changed", '']  # Initialize lines.
+    for commit_type, items in categories.items():
+        if items:
+            lines.append(f'### {commit_type_to_category(commit_type)}')  # Add category header.
+            for description, pr_link in items:
+                lines.append(f'* {description} in {pr_link}')  # Add commit description.
+            lines.append('')  # Add a blank line after each category.
+    if full_changelog:
+        lines.append(full_changelog)  # Append the full changelog.
+    return '\n'.join(lines)  # Join the lines with newline characters.
```
**Minor:** consider removing these comments, it's pretty clear what the code does.
